src/model.rs

```diff
@@ -0,0 +1,118 @@
+use core::f32;
+use std::{collections::BTreeMap, error::Error, path::PathBuf};
+
+use serde::{Deserialize, Serialize};
+use sha2::{Digest, Sha256};
+
+use crate::{error::SerializeError, utils::get_struct_name};
+
+pub fn to_yaml<T: Serialize + std::fmt::Debug>(item: T) -> Result<String, Box<dyn Error>> {
+    let mut yaml_str = match serde_yaml::to_string(&item) {
+        Ok(value) => value,
+        Err(error) => {
+            return Err(Box::new(SerializeError {
+                item_debug_string: format!("{:?}", item),
+                error,
+            }))
+        }
+    };
+
+    yaml_str.insert_str(0, &format!("class: {}\n", get_struct_name::<T>()));
+
+    Ok(yaml_str)
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Annotation {
+    pub name: String,
+    pub description: String,
+    pub version: String,
+}
+
+/// String would be the name of the model
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
+pub enum GPUVendorInfo {
+    NVIDIA(String),
+    AMD(String),
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
+pub struct GPUSpecRequirement {
+    pub vendor_info: GPUVendorInfo,
+    pub min_memory: u64,
+    pub count: u16,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
+pub struct KeyInfo {
+    pub path: PathBuf,
+    pub matching_pattern: String,
+}
+
+#[derive(Serialize, Debug)]
+pub struct Pod {
+    #[serde(skip_serializing)]
+    pub annotation: Annotation,
+    pub gpu_spec_requirments: Option<GPUSpecRequirement>,
+    pub image_digest: String,
+    pub input_stream_map: BTreeMap<String, KeyInfo>, // Num of recommneded cpu cores (can be fractional)
+    pub min_memory: u64,
+    pub output_dir: PathBuf,
+    pub output_stream_map: BTreeMap<String, KeyInfo>,
+    #[serde(skip_serializing)]
+    pub pod_hash: String, // SHA256 docker image hash
+    pub recommended_cpus: f32,
+    pub source_commit: String, // Git Commit
+}
+
+#[derive(Clone)]
+pub struct PodNewConfig {
+    pub name: String,
+    pub description: String,
+    pub version: String,
+    pub input_stream_map: BTreeMap<String, KeyInfo>,
+    pub output_dir: PathBuf,
+    pub output_stream_map: BTreeMap<String, KeyInfo>, // Will be relative path of output_dir
+    pub image_name: String,
+    pub source_commit: String,         // Git Commit only for reference
+    pub recommended_cpus: Option<f32>, // Num of recommneded cpu cores (can be fractional)
+    pub min_memory: Option<u64>,       // Bytes
+    pub gpu_spec_requirments: Option<GPUSpecRequirement>,
+}
```

Comment:
Not a big fan of this approach since it still feels quite similar to explicitly defining all of these in a `new(...)` but with added confusion since user would need to interact with both and it makes the codebase a bit harder to follow as the models grow. Since the default is still set in `Pod::new` anyway, I'd prefer to just defining them there. Have a look how I approached this [here](https://github.com/walkerlab/orcapod/pull/5/files#diff-6bd13171e8498a8adc17d13cef1f1b59f5fbd24ae4c72aeb4ce5f9b2b4703220R71-R109).

Reply:
merge into new TODO